packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,323 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to define the field to sort by. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Spec](./README.md)
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text data-key="name"></nimble-table-column-text>
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the columns elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: TableColumn[] = [];
```

Comment:
nit: Why include "slotted" in the name? Is that a naming convention we typically use or could this be simplified to just `columns` or `columnDefinitions`?

Related: if the user sets this property does it automatically update the state of the DOM? Or should this be made `readonly` with the only API to modify it going through DOM APIs like add/removeElement?

If there are FAST patterns I should educate myself about that already explain all this, feel free to let me know.

Reply:
Sorry this should be marked as `internal`. I was following a pattern put in place in the [`menu-button`](https://github.com/ni/nimble/blob/f3692b7db86a15b4bc0f00151a736a5f9de938df/packages/nimble-components/src/menu-button/index.ts#L58) for its slotted menus. Perhaps there is a good reason to not name it this? Ultimately, I'm not too concerned about it, because, again, this is expected to be an internal API (but you are correct in that it should be marked as `readonly`).